### PR TITLE
r/secretsmanager_secret_policy - new resource

### LINKED ISSUE #14618: New resource aws_secretsmanager_resource_policy
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
It would be interesting to have the ability to decouple the creation of a Secrets Manager from the definition of its resource policy. This is useful if you need to "inject" the policy from a separate module, especially when that happens after the secret already exists.

### New or Affected Resource(s)

* aws_secretsmanager_resource_policy

### Potential Terraform Configuration

```hcl
resource "aws_secretsmanager_secret" "example" {
  name = "example"
}

resource "aws_secretsmanager_resource_policy" "example_policy {
  secret_arn = aws_secretsmanager_secret.example.arn
  block_public_policy = true
  policy = <JSON policy>
}
```

### References

* [PutResourcePolicy](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_PutResourcePolicy.html)
* [DeleteResourcePolicy](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_DeleteResourcePolicy.html)

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #14618
Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_secretsmanager_secret_policy - new resource
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAwsSecretsManagerSecretPolicy_'
--- PASS: TestAccAwsSecretsManagerSecretPolicy_basic (92.55s)
--- PASS: TestAccAwsSecretsManagerSecretPolicy_blockPublicPolicy (122.51s)
--- PASS: TestAccAwsSecretsManagerSecretPolicy_disappears (48.91s)
```

### CODE REVIEW

**Comment by DrFaust92** [comment]
Rebased + added support for blocking public access:

```
--- PASS: TestAccAwsSecretsManagerSecretPolicy_basic (92.55s)
--- PASS: TestAccAwsSecretsManagerSecretPolicy_blockPublicPolicy (122.51s)
--- PASS: TestAccAwsSecretsManagerSecretPolicy_disappears (48.91s)
```

**Comment by haidaraM** [comment]
Hi, 
Thanks for this PR :-) Any idea of when this will be merged.

**Comment by dblooman** [comment]
Can we get an update on this PR please 👍

File: aws/resource_aws_secretsmanager_secret_policy.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+			"This resource policy contains an unsupported principal") {
+			return resource.RetryableError(err)
+		}
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if isResourceTimeoutError(err) {
+		res, err = conn.PutResourcePolicy(input)
```
Comment (bill-rich) [inline aws/resource_aws_secretsmanager_secret_policy.go:75]: Why is the policy being applied again if the previous attempt timed out? Is this an expected condition?

File: aws/resource_aws_secretsmanager_secret_policy.go [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+				"This resource policy contains an unsupported principal") {
+				return resource.RetryableError(err)
+			}
+			if err != nil {
+				return resource.NonRetryableError(err)
+			}
+			return nil
+		})
+		if isResourceTimeoutError(err) {
+			_, err = conn.PutResourcePolicy(input)
```
Comment (bill-rich) [inline aws/resource_aws_secretsmanager_secret_policy.go:145]: Same as above

**Review by bill-rich — COMMENTED** [review]
Looks great overall. I had about one of the conditions that came up twice. Please let me know on those.

File: aws/resource_aws_secretsmanager_secret_policy.go (hunk 1 above)
Comment (bflad) [reply aws/resource_aws_secretsmanager_secret_policy.go:75]: @bill-rich this is required in the AWS Provider due to how the AWS Go SDK works, see also:

* https://github.com/terraform-providers/terraform-provider-aws/issues/7873
* https://github.com/terraform-providers/terraform-provider-aws/issues/12985

**Comment by DrFaust92** [comment]
Rebased for acceptance test config lint

**Review by bill-rich — APPROVED** [review]
I got my question answered. LGTM

**Pull request merged**